storm-core/src/jvm/org/apache/storm/daemon/supervisor/ReadClusterState.java

```diff
@@ -0,0 +1,318 @@
+        Map<String, VersionedData<Assignment>> updateAssignmentVersion = new HashMap<>();
+        for (String topoId : topoIds) {
+            Integer recordedVersion = -1;
+            Integer version = stormClusterState.assignmentVersion(topoId, callback);
+            VersionedData<Assignment> locAssignment = localAssignmentVersion.get(topoId);
+            if (locAssignment != null) {
+                recordedVersion = locAssignment.getVersion();
+            }
+            if (version == null) {
+                // ignore
+            } else if (version == recordedVersion) {
+                updateAssignmentVersion.put(topoId, locAssignment);
+            } else {
+                VersionedData<Assignment> assignmentVersion = stormClusterState.assignmentInfoWithVersion(topoId, callback);
+                updateAssignmentVersion.put(topoId, assignmentVersion);
+            }
+        }
+        return updateAssignmentVersion;
+    }
+    
+    protected Map<String, List<ProfileRequest>> getProfileActions(IStormClusterState stormClusterState, List<String> stormIds) throws Exception {
+        Map<String, List<ProfileRequest>> ret = new HashMap<String, List<ProfileRequest>>();
+        for (String stormId : stormIds) {
+            List<ProfileRequest> profileRequests = stormClusterState.getTopologyProfileRequests(stormId);
+            ret.put(stormId, profileRequests);
+        }
+        return ret;
+    }
+    
+    protected Map<Integer, LocalAssignment> readAssignments(Map<String, VersionedData<Assignment>> assignmentsSnapshot,
+            String assignmentId, AtomicInteger retries) {
+        try {
+            Map<Integer, LocalAssignment> portLA = new HashMap<Integer, LocalAssignment>();
+            for (Map.Entry<String, VersionedData<Assignment>> assignEntry : assignmentsSnapshot.entrySet()) {
+                String topoId = assignEntry.getKey();
+                Assignment assignment = assignEntry.getValue().getData();
+
+                Map<Integer, LocalAssignment> portTasks = readMyExecutors(topoId, assignmentId, assignment);
+
+                for (Map.Entry<Integer, LocalAssignment> entry : portTasks.entrySet()) {
+
+                    Integer port = entry.getKey();
+
+                    LocalAssignment la = entry.getValue();
+
+                    if (!portLA.containsKey(port)) {
+                        portLA.put(port, la);
+                    } else {
+                        throw new RuntimeException("Should not have multiple topologies assigned to one port");
+                    }
+                }
+            }
+            retries.set(0);
+            return portLA;
+        } catch (RuntimeException e) {
+            if (retries.get() > 2) {
+                throw e;
+            } else {
+                retries.addAndGet(1);
+            }
+            LOG.warn("{} : retrying {} of 3", e.getMessage(), retries.get());
```

Comment:
This is code that came from clojure.  It might be worth rewriting the entire function to not use the AtomicInteger, but I thought it would be best to leave the code alone, and if we wanted to clean it up later we could.